pkg/operator/controller/controller_dns.go

```diff
@@ -1,89 +1,135 @@
+		return nil, err
 	}
-}
-
-func newARecord(domain, target string, zone configv1.DNSZone) *dns.Record {
-	return &dns.Record{
-		Zone: zone,
-		Type: dns.ARecordType,
-		ARecord: &dns.ARecord{
-			Domain:  domain,
-			Address: target,
-		},
+	if desired != nil && current == nil {
+		if err := r.client.Create(context.TODO(), desired); err != nil {
+			return nil, fmt.Errorf("failed to create dnsrecord %s/%s: %v", desired.Namespace, desired.Name, err)
+		}
+		log.Info("created dnsrecord", "namespace", desired.Namespace, "name", desired.Name)
+		return desired, nil
 	}
+	return current, nil
 }
 
-// desiredDNSRecords will return any necessary DNS records for the given inputs.
-// If an ingress domain is in use, records are desired in every specified zone
-// present in the cluster DNS configuration.
-func desiredDNSRecords(ci *operatorv1.IngressController, dnsConfig *configv1.DNS, service *corev1.Service) []*dns.Record {
-	records := []*dns.Record{}
-
+// desiredDNSRecords will return any necessary wildcard DNS records for the
+// ingresscontroller.
+//
+// For now, if the service has more than one .status.loadbalancer.ingress, only
+// the first will be used.
+func desiredWildcardRecord(ic *operatorv1.IngressController, service *corev1.Service) *iov1.DNSRecord {
 	// If the ingresscontroller has no ingress domain, we cannot configure any
 	// DNS records.
-	if len(ci.Status.Domain) == 0 {
-		return records
+	if len(ic.Status.Domain) == 0 {
+		return nil
 	}
 
-	// If the HA type is not cloud, then we don't manage DNS.
-	if ci.Status.EndpointPublishingStrategy.Type != operatorv1.LoadBalancerServiceStrategyType {
-		return records
+	// DNS is only managed for LB publishing.
+	if ic.Status.EndpointPublishingStrategy.Type != operatorv1.LoadBalancerServiceStrategyType {
+		return nil
 	}
 
-	name := fmt.Sprintf("*.%s", ci.Status.Domain)
-	zones := []configv1.DNSZone{}
-	if dnsConfig.Spec.PrivateZone != nil {
-		zones = append(zones, *dnsConfig.Spec.PrivateZone)
+	// No LB target exists for the domain record to point at.
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return nil
 	}
-	if dnsConfig.Spec.PublicZone != nil {
-		zones = append(zones, *dnsConfig.Spec.PublicZone)
+
+	ingress := service.Status.LoadBalancer.Ingress[0]
+
+	// Quick sanity check since we don't know how to handle both being set (is
+	// that even a valid state?)
+	if len(ingress.Hostname) > 0 && len(ingress.IP) > 0 {
+		return nil
 	}
-	for _, ingress := range service.Status.LoadBalancer.Ingress {
-		if len(ingress.Hostname) > 0 {
-			for _, zone := range zones {
-				records = append(records, newAliasRecord(name, ingress.Hostname, zone))
-			}
-		}
-		if len(ingress.IP) > 0 {
-			for _, zone := range zones {
-				records = append(records, newARecord(name, ingress.IP, zone))
-			}
+
+	name := WildcardDNSRecordName(ic)
+	domain := fmt.Sprintf("*.%s", ic.Status.Domain)
+	var target string
+	var recordType string
+
+	if len(ingress.Hostname) > 0 {
+		recordType = "ALIAS"
```

Comment:
"ALIAS" is AWS-specific.  Let's document that and think about where the logic to choose between "ALIAS" or "CNAME" belongs.

Reply:
Added some comments